### ISSUE TITLE
REQ_WCAG_1.1.1 Non-text Content - Situation C: If non-text content is a control or accepts user input

### ISSUE THREAD

**Issue by ecureuill** [solex2006/SELIProject#16]
All non-text content that is presented to the user has a text alternative that serves the equivalent purpose

## Intent
Text alternatives are a primary way for making information accessible because they can be rendered through any sensory modality (for example, visual, auditory or tactile) to match the needs of the user. Providing text alternatives allows the information to be rendered in a variety of ways by a variety of user agents

## Sufficient Techniques

> Sufficient techniques are reliable ways to meet the success criteria.
> - From an author's perspective: If you use the sufficient techniques for a given criterion correctly and it is accessibility-supported for your users, you can be confident that you met the success criterion.
> - From an evaluator's perspective: If web content implements the sufficient techniques for a given criterion correctly and it is accessibility-supported for the content's users, it conforms to that success criterion. (The converse is not true; if content does not implement these sufficient techniques, it does not necessarily fail the success criteria, as explained in Testing Techniques below.)
> There may be other ways to meet success criteria besides the sufficient techniques in W3C's Techniques for WCAG document, as Other Techniques below. (See also Techniques are Informative above.)

### [G82: Providing a text alternative that identifies the purpose of the non-text content ](https://www.w3.org/WAI/WCAG21/Techniques/general/G82) using one of the following Text alternative techniques:
>  provide useful information via the text alternative even if the full function of the non-text content cannot be provided.
-   [ARIA6: Using aria-label to provide labels for **objects**](https://www.w3.org/WAI/WCAG21/Techniques/aria/ARIA6)
-   [ARIA9: Using aria-labelledby to concatenate a label from **several text nodes**](https://www.w3.org/WAI/WCAG21/Techniques/aria/ARIA9)
-   [H24: Providing text alternatives for the area elements of **image maps**](https://www.w3.org/WAI/WCAG21/Techniques/html/H24)
-   [H30: Providing link text that describes the purpose of a link for **anchor elements**](https://www.w3.org/WAI/WCAG21/Techniques/html/H30)
-   [H36: Using alt attributes on images used as submit buttons](https://www.w3.org/WAI/WCAG21/Techniques/html/H36)
-   [H44: Using label elements to associate text labels with **form controls**](https://www.w3.org/WAI/WCAG21/Techniques/html/H44)
-   [H65: Using the title attribute to identify **form controls** when the label element cannot be used](https://www.w3.org/WAI/WCAG21/Techniques/html/H65)

#### Test Procedure
G82
1. remove, hide, or mask the non-text content
2. replace it with the text alternative
3. check that the purpose of the non-text content is clear - even if function is lost.

**Expected Results**
Check # 3 is true.

# SELI CONTEXT

## Tasks
List the tasks required to accomplish the requirement (**not exhaustive**)

- [ ] **SYSTEM** MUST have a text alternative for each control and user input that Student should interact. 

- [ ] **TEACHER** MUST have the option to include text alternative for link content.

# :notebook_with_decorative_cover: References

- [How to Meet WCAG (Quick Reference)](https://www.w3.org/WAI/WCAG21/quickref)
- [Understanding Techniques for WCAG Success Criteria](https://www.w3.org/WAI/WCAG21/Understanding/understanding-techniques)

# GLOSSARY 
- SELI's content: all contents not created by users (e.g. seli "about" section text)

:warning: The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED",  "MAY", and "OPTIONAL" in this document are to be interpreted as described in [RFC 2119/8174](http://www.rfc-editor.org/info/rfc8174).
> 1. **MUST** This word, or the terms "REQUIRED" or "SHALL", mean that the definition is an absolute requirement of the specification.
> 2. **MUST NOT** This phrase, or the phrase "SHALL NOT", mean that the definition is an absolute prohibition of the specification.
> 3. **SHOULD** This word, or the adjective "RECOMMENDED", mean that there may exist valid reasons in particular circumstances to ignore a particular item, but the full implications must be understood and carefully weighed before choosing a different course.
> 4. **SHOULD NOT** This phrase, or the phrase "NOT RECOMMENDED" mean that there may exist valid reasons in particular circumstances when the particular behavior is acceptable or even useful, but the full implications should be understood and the case carefully weighed before implementing any behavior described with this label.
> 5. **MAY**   This word, or the adjective "OPTIONAL", mean that an item is truly optional. An implementation which does not include a particular option MUST be prepared to interoperate with another implementation which does include the option, though perhaps with reduced functionality. In the same vein an implementation which does include a particular option MUST be prepared to interoperate with another implementation which does not include the option (except, of course, for the feature the option provides.)

**Comment by juantenesaca** [comment]
In this aria-labels, ids and labels, do I have to use the variables of props.language or is it ok to just use English on everything? or just for aria-labels.

**Comment by ecureuill** [comment]
labels, aria-label, aria-describedby, etc should be with props.language.

id just need to be unique, so for reusable components create an id with ` {some_prefix}+{id}`
https://lodash.com/docs#uniqueId

```js
import React, { useState } from 'react'
import uniqueId from 'lodash/utility/uniqueId'

const Field = props => {
  const [ id ] = useState(() => uniqueId('myprefix-'))

  return (
    <div>
      <label htmlFor={id}>{props.label}</label>
      <input id={id} type="text"/>
    </div>
  )      
}

export default Field
```

**Comment by juantenesaca** [comment]
OK that is good. But, for now let's left it until new features of instructional design were completely implemented. Because if we start now with it. Then you have to review again.

**Comment by ecureuill** [comment]
@juantenesaca  when you finish, just move the card to CODED list.
![image](https://user-images.githubusercontent.com/993369/79517432-8793a680-8024-11ea-9da4-3752a3bfb3d6.png)

The **2 - Ready**, means it' ready to be coded (after specification phase).

**Comment by ecureuill** [comment]
# Failures

## Course Topic
- [ ] due to breadcrumb
- [ ] due to missing Accessible Name or Accessible Description for Course Structure buttons
![image](https://user-images.githubusercontent.com/993369/92221435-da56e900-ee73-11ea-9b77-e0d164fcedd2.png)
- [ ] due to close button at course menu
![image](https://user-images.githubusercontent.com/993369/91644232-947fc800-ea10-11ea-8c61-acc8c6db5d76.png)

# Course Content

- [ ] due to missing Accessible Name or Accessible Description for audio component 
![image](https://user-images.githubusercontent.com/993369/92087272-ae6b3300-eda1-11ea-82df-faf6d70de9a4.png)
- [ ] due to frames without title
- [ ] due to emulating link instead of use `<a>` in Link component
